### ISSUE TITLE
Error: status == CUDNN_STATUS_SUCCESS (8 vs. 0)  CUDNN_STATUS_EXECUTION_FAILED

### ISSUE THREAD

**Issue by alfredox10** [NVIDIA/DIGITS#232]
I am getting this error when trying to run training with my custom network.

status == CUDNN_STATUS_SUCCESS (8 vs. 0)  CUDNN_STATUS_EXECUTION_FAILED

I found this post that refers to this error:
https://github.com/BVLC/caffe/issues/1700#issuecomment-133476490

But it doesn't specify where or how to fix it. Also I am not sure if the issues are related or something completely different. Let me mention that this custom framework works perfectly fine when I run it in my local caffe install, and I can also see all the nodes if I hit the visualize button. It starts training and fails after the first epoch.

![pasted_image_at_2015_08_21_12_18_am](https://cloud.githubusercontent.com/assets/11828765/9416372/e88e8a74-480a-11e5-88ab-4721d91b7254.png)

**Comment by lukeyeager** [comment]
Hi @alfredox10, can you tell me which versions you are using of DIGITS, Caffe, cuDNN and CUDA?

Also, can you show me the last dozen lines or so of your raw Caffe output? That may provide a stacktrace which can help us zero in on the issue.

**Comment by alfredox10** [comment]
Digits-2.0rc, CUDA 7.0, Caffe is coming from inside DIGITS which is the 0.13 NVIDIA branch, CuDNN v1

Last few lines of caffe_output.log:
I0821 00:16:38.837415 31534 solver.cpp:217] Iteration 2291, loss = 0.0492331
I0821 00:16:38.837440 31534 solver.cpp:234]     Train net output #0: loss1 = 0.0492332 (* 1 = 0.0492332 loss)
I0821 00:16:38.837469 31534 solver.cpp:511] Iteration 2291 (21.5399/s), lr = 0.01
I0821 00:16:42.570436 31534 solver.cpp:217] Iteration 2370, loss = 0.0741715
I0821 00:16:42.570472 31534 solver.cpp:234]     Train net output #0: loss1 = 0.0741716 (* 1 = 0.0741716 loss)
I0821 00:16:42.570519 31534 solver.cpp:511] Iteration 2370 (21.1626/s), lr = 0.01
I0821 00:16:46.272135 31534 solver.cpp:217] Iteration 2449, loss = 0.0256392
I0821 00:16:46.272162 31534 solver.cpp:234]     Train net output #0: loss1 = 0.0256393 (* 1 = 0.0256393 loss)
I0821 00:16:46.272452 31534 solver.cpp:511] Iteration 2449 (21.3406/s), lr = 0.01
F0821 00:16:49.554188 31534 cudnn_conv_layer.cu:126] Check failed: status == CUDNN_STATUS_SUCCESS (8 vs. 0)  CUDNN_STATUS_EXECUTION_FAILED
*** Check failure stack trace: ***
@     0x7fb302368daa  (unknown)
@     0x7fb302368ce4  (unknown)
@     0x7fb3023686e6  (unknown)
@     0x7fb30236b687  (unknown)
@     0x7fb302a432e1  caffe::CuDNNConvolutionLayer<>::Backward_gpu()
@     0x7fb302a2b367  caffe::Net<>::BackwardFromTo()
@     0x7fb302a2b4d1  caffe::Net<>::Backward()
@     0x7fb302a177f6  caffe::Solver<>::Step()
@     0x7fb302a182ee  caffe::Solver<>::Solve()
@           0x4078a8  train()
@           0x405421  main
@     0x7fb30187aec5  (unknown)
@           0x405a4b  (unknown)
@              (nil)  (unknown)

**Comment by lukeyeager** [comment]
Oh ok, this is from the web installer. cuDNN version is actually v3.

@slayton58, can you take a look at this?

**Comment by lukeyeager** [comment]
> Digits-2.0rc, CUDA 7.0, Caffe is coming from inside DIGITS which is the 0.13 NVIDIA branch, CuDNN v1

What do you mean by "Caffe is coming from inside DIGITS"? Did you build Caffe yourself with [these instructions](https://github.com/NVIDIA/DIGITS/blob/digits-2.0/docs/InstallCaffe.md), or did you use the [web installer](https://developer.nvidia.com/digits)?

If you did build from source, Caffe v0.13 expects cuDNN v3. Maybe that's your problem?
https://github.com/NVIDIA/caffe/wiki/Versions

**Comment by alfredox10** [comment]
I built it via the web installer, so there is a Caffe folder inside the DIGITS folder.

**Comment by lukeyeager** [comment]
@alfredox10, can you send me your custom network? I can't reproduce this.

**Comment by alfredox10** [comment]
Yes this is what it looks like:

name: "LogisticRegressionNet"
layers {
  name: "data"
  type: HDF5_DATA
  top: "data"
  top: "label"
  hdf5_data_param {
    source: "examples/bibsmart/train.txt"
    batch_size: 64
  }
  transform_param {
    scale: 0.005
  }
  include: { phase: TRAIN }
}
layers {
  name: "data"
  type: HDF5_DATA
  top: "data"
  top: "label"
  hdf5_data_param {
    source: "examples/bibsmart/test.txt"
    batch_size: 64
  }
  transform_param {
    scale: 0.005
  }
  include: { phase: TEST }
}
layers {
  name: "conv1"
  type: CONVOLUTION
  bottom: "data"
  top: "conv1"
  blobs_lr: 1
  blobs_lr: 2
  convolution_param {
    num_output: 48
    pad: 0
    kernel_size: 5
    stride: 1
    weight_filler {
      type: "xavier"
    }
    bias_filler {
      type: "constant"
    }
  }
}
layers {
  name: "drop1"
  type: DROPOUT
  bottom: "conv1"
  top: "conv1"
  dropout_param {
    dropout_ratio: 0
  }
}
layers {
  name: "relu1"
  type: RELU
  bottom: "conv1"
  top: "conv1"
}
layers {
  name: "norm1"
  type: LRN
  bottom: "conv1"
  top: "norm1"
  lrn_param {
    local_size: 3
    alpha: 0.0001
    beta: 0.75
  }
}
layers {
  name: "pool1"
  type: POOLING
  bottom: "norm1"
  top: "pool1"
  pooling_param {
    pool: MAX
    kernel_size: 2
    stride: 2
  }
}
layers {
  name: "conv2"
  type: CONVOLUTION
  bottom: "pool1"
  top: "conv2"
  blobs_lr: 1
  blobs_lr: 2
  convolution_param {
    num_output: 64
    pad: 2
    kernel_size: 5
    stride: 1
    weight_filler {
      type: "xavier"
    }
    bias_filler {
      type: "constant"
    }
  }
}
layers {
  name: "drop2"
  type: DROPOUT
  bottom: "conv2"
  top: "conv2"
  dropout_param {
    dropout_ratio: 0
  }
}
layers {
  name: "relu2"
  type: RELU
  bottom: "conv2"
  top: "conv2"
}
layers {
  name: "norm2"
  type: LRN
  bottom: "conv2"
  top: "norm2"
  lrn_param {
    local_size: 3
    alpha: 0.0001
    beta: 0.75
  }
}
layers {
  name: "pool2"
  type: POOLING
  bottom: "norm2"
  top: "pool2"
  pooling_param {
    pool: MAX
    kernel_size: 2
    stride: 1
  }
}
layers {
  name: "conv3"
  type: CONVOLUTION
  bottom: "pool2"
  top: "conv3"
  blobs_lr: 1
  blobs_lr: 2
  convolution_param {
    num_output: 128
    pad: 1
    kernel_size: 5
    stride: 1
    weight_filler {
      type: "xavier"
    }
    bias_filler {
      type: "constant"
    }
  }
}
layers {
  name: "drop3"
  type: DROPOUT
  bottom: "conv3"
  top: "conv3"
  dropout_param {
    dropout_ratio: 0
  }
}
layers {
  name: "relu3"
  type: RELU
  bottom: "conv3"
  top: "conv3"
}
layers {
  name: "norm3"
  type: LRN
  bottom: "conv3"
  top: "norm3"
  lrn_param {
    local_size: 3
    alpha: 0.0001
    beta: 0.75
  }
}
layers {
  name: "pool3"
  type: POOLING
  bottom: "norm3"
  top: "pool3"
  pooling_param {
    pool: MAX
    kernel_size: 2
    stride: 2
  }
}
layers {
  name: "conv4"
  type: CONVOLUTION
  bottom: "pool3"
  top: "conv4"
  blobs_lr: 1
  blobs_lr: 2
  convolution_param {
    num_output: 128
    pad: 1
    kernel_size: 5
    stride: 1
    weight_filler {
      type: "xavier"
    }
    bias_filler {
      type: "constant"
    }
  }
}
layers {
  name: "drop4"
  type: DROPOUT
  bottom: "conv4"
  top: "conv4"
  dropout_param {
    dropout_ratio: 0.5
  }
}
layers {
  name: "relu4"
  type: RELU
  bottom: "conv4"
  top: "conv4"
}
layers {
  name: "norm4"
  type: LRN
  bottom: "conv4"
  top: "norm4"
  lrn_param {
    local_size: 3
    alpha: 0.0001
    beta: 0.75
  }
}
layers {
  name: "pool4"
  type: POOLING
  bottom: "norm4"
  top: "pool4"
  pooling_param {
    pool: MAX
    kernel_size: 2
    stride: 2
  }
}
layers {
  name: "conv5"
  type: CONVOLUTION
  bottom: "pool4"
  top: "conv5"
  blobs_lr: 1
  blobs_lr: 2
  convolution_param {
    num_output: 128
    pad: 1
    kernel_size: 5
    stride: 1
    weight_filler {
      type: "xavier"
    }
    bias_filler {
      type: "constant"
    }
  }
}
layers {
  name: "drop5"
  type: DROPOUT
  bottom: "conv5"
  top: "conv5"
  dropout_param {
    dropout_ratio: 0.5
  }
}
layers {
  name: "relu5"
  type: RELU
  bottom: "conv5"
  top: "conv5"
}
layers {
  name: "norm5"
  type: LRN
  bottom: "conv5"
  top: "norm5"
  lrn_param {
    local_size: 3
    alpha: 0.0001
    beta: 0.75
  }
}
layers {
  name: "pool5"
  type: POOLING
  bottom: "norm5"
  top: "pool5"
  pooling_param {
    pool: MAX
    kernel_size: 2
    stride: 2
  }
}
layers {
  name: "ip1"
  type: INNER_PRODUCT
  bottom: "pool5"
  top: "ip1"
  blobs_lr: 1
  blobs_lr: 2
  inner_product_param {
    num_output: 3072
    weight_filler {
      type: "xavier"
    }
    bias_filler {
      type: "constant"
    }
  }
}
layers {
  name: "relu9"
  type: RELU
  bottom: "ip1"
  top: "ip1"
}
layers {
  name: "drop9"
  type: DROPOUT
  bottom: "ip1"
  top: "ip1"
  dropout_param {
    dropout_ratio: 0.5
  }
}
layers {
  name: "ip2"
  type: INNER_PRODUCT
  bottom: "ip1"
  top: "ip2"
  blobs_lr: 1
  blobs_lr: 2
  inner_product_param {
    num_output: 2
    weight_filler {
      type: "xavier"
    }
    bias_filler {
      type: "constant"
    }
  }
}
layers {
  name: "accuracy1"
  type: ACCURACY
  bottom: "ip2"
  bottom: "label"
  top: "accuracy1"
  include: { phase: TEST }
}
layers {
  name: "loss1"
  type: SOFTMAX_LOSS
  bottom: "ip2"
  bottom: "label"
  top: "loss1"
}

**Comment by lukeyeager** [comment]
@alfredox10, sorry about the delay. This bug got lost in the pile.

I'm running your network and still not seeing any issues. I'm using 256x256 images - should I be using something else?

Do you get the same errors when using one of the standard networks?

**Comment by alfredox10** [comment]
I am using 64x64 images. I do not get this error with the standard networks.

**Comment by lukeyeager** [comment]
Hmm, still no problems with 64x64 images on my machine.

Which GPU[s] do you have?

**Comment by alfredox10** [comment]
I use a 780 Ti. Ok well let me try this again soon with the same network
and with the standard ones. I just upgraded my machine to nvidia driver
352, CUDA 7.5, Caffe 0.13 nvidia branch, and CuDNN v3, and I'll see if i'm
still having those same issues.




*Jose Alfredo NavaUniversity of Houston - Clear Lake | Computer Engineering*
*Institute of Electrical and Electronics Engineers*

*http://ainavbots.com/ <http://ainavbots.com/>*

**Comment by lukeyeager** [comment]
@alfredox10 did you resolve your issue?

**Comment by alfredox10** [comment]
No I was never able to solve it, I went back to using Caffe by itself for training. Going to try to figure out if I can run multiple GPU training using Caffe 0.13, is there a guide on how to set it up to do that?

**Comment by lukeyeager** [comment]
You don't really have to set up anything. Just [install Caffe](https://github.com/NVIDIA/DIGITS/blob/v2.2.1/docs/InstallCaffe.md) and select more than one GPU when creating your model:

![multi-gpu-selection](https://cloud.githubusercontent.com/assets/687269/10584416/04769140-7647-11e5-87cc-a3eb9829d40e.jpg)

**Comment by alfredox10** [comment]
Well this is great, but I can't use the caffemodel from DIGITS because it doesn't output the same data structure as the caffemodel produced when manually training with Caffe from the terminal. So I'm just using Caffe by itself, I meant what is the additional command needed to run Caffe from the terminal in multi-GPU way? Right now I'm running it like this due to some bug that doesn't set to GPU 0 if not specified.

build/tools/caffe train -gpu 0 --solver=examples/bibsmart/bibsmart_solver.prototxt

**Comment by lukeyeager** [comment]
> I can't use the caffemodel from DIGITS because it doesn't output the same data structure as the caffemodel produced when manually training with Caffe from the terminal.

It should. Can you explain the difference you're seeing?

> some bug that doesn't set to GPU 0 if not specified.

Same as this issue? https://github.com/NVIDIA/caffe/issues/23

**Comment by thatguymike** [comment]
For manually running multiple GPUs, list the device ordinals you want to use, e.g. instead of 0 you can use 0,1

**Comment by lukeyeager** [comment]
```
$ build/tools/caffe --help
caffe: command line brew
usage: caffe <command> <args>

commands:
  train           train or finetune a model
  test            score a model
  device_query    show GPU diagnostic information
  time            benchmark model execution time

  Flags from /home/lyeager/caffe/nv-0.14.0-alpha/tools/caffe.cpp:
    -gpu (Optional; run in GPU mode on given device IDs separated by ','.Use
      '-gpu all' to run on all available GPUs. The effective training batch
      size is multiplied by the number of devices.) type: string default: ""
...
```

**Comment by alfredox10** [comment]
Yes same as issue NVIDIA/caffe#23, however when I made those changes to the code in the 0.13 branch, it returned type errors, so I just set it manually and that fixed it. Ok thanks guys! I'll try it and let ya'll know if that works. Once I test that I may try to reuse the DIGITS caffemodel output, I'll come back here and show you guys what it's producing, because it does not give me the same predict model with 'ip2' and 'labels' as keys.

**Comment by alfredox10** [comment]
I received two errors when trying to run the -gpu all and the -gpu 0,1 command.

alfredo@alfpc-ubuntu:~/Dropbox/Alfredo1/Sodeypunk_repo_alfredo/caffe-dev-0.13$ build/tools/caffe train -gpu all --solver=examples/bibsmart/bibsmart_solver.prototxt
ERROR: illegal value 'all' specified for int32 flag 'gpu'
alfredo@alfpc-ubuntu:~/Dropbox/Alfredo1/Sodeypunk_repo_alfredo/caffe-dev-0.13$ build/tools/caffe train -gpu 0,1 --solver=examples/bibsmart/bibsmart_solver.prototxt
ERROR: illegal value '0,1' specified for int32 flag 'gpu'

Also how do you enable multi-gpu when you are using the train file to perform detections? I know there is the caffe.set_mode_gpu() command, is there another to enable both GPUs?

**Comment by lukeyeager** [comment]
Sorry, I gave you the wrong information before. With NVcaffe v0.12 and v0.13, [we went with the flags](https://github.com/NVIDIA/caffe/blob/v0.13.2/tools/caffe.cpp#L23-L27) proposed by @cypof in his original multi-GPU pull request (https://github.com/BVLC/caffe/pull/2114):
```
DEFINE_int32(gpu, -1,
    "Run in GPU mode on given device ID (Legacy switch, use -gpus).");
DEFINE_string(gpus, "",
    "Run in GPU mode on given device IDs separated by ','."
    "Use '-gpus all' to run on all available GPUs.");
```
With v0.14, [we are switching to the flags](https://github.com/NVIDIA/caffe/blob/v0.14.0-beta/tools/caffe.cpp#L31-L34) that BVLC actually merged into their fork:
```
DEFINE_string(gpu, "",
    "Optional; run in GPU mode on given device IDs separated by ','."
    "Use '-gpu all' to run on all available GPUs. The effective training "
    "batch size is multiplied by the number of devices.");
```
What you want is `--gpus=0,1`, not `--gpu=0,1`.

Again, you should be able to see this for yourself with `caffe --help`.

**Comment by lukeyeager** [comment]
> Also how do you enable multi-gpu when you are using the train file to perform detections? I know there is the caffe.set_mode_gpu() command, is there another to enable both GPUs?

It looks like you've added a new issue for this at https://github.com/NVIDIA/caffe/issues/43. I don't actually know the answer to that - I'll let someone else respond in that thread.

**Comment by lukeyeager** [comment]
@alfredox10 there are a lot of divergent issues in this thread. I'm going to close this issue. Please re-open a new issue for any remaining against DIGITS or Caffe as appropriate.

**Comment by alfredox10** [comment]
Hey Luke, your suggestion for setting the flags with -gpus 0,1 worked.

Thanks
